proposals/0054-sessions.md

```diff
@@ -0,0 +1,219 @@
+
+# Goals
+
+- Simple key/value API, similar to Astro cookies (`Astro.session.get('key')`, `Astro.session.set('key', 'value')`)
+- Accessible in Astro components, actions, and API endpoints via Astro global and context objects.
+- Additional control over the session via methods such as `Astro.session.regenerate()` and `Astro.session.destroy()`
+- Storage drivers for popular providers available out of the box.
+- Adapters can specify a default storage driver, with Node using filesystem storage by default.
+- Other settings such as cookie name, session expiry etc are also optionally configurable.
+- Sessions are lazy-loaded and auto-generated when first accessed. Session ID cookie is get and set automatically.
+
+# Non-Goals
+
+- Type-safety. This may be added later, but there are questions about the best way to do this, because of issues with the way that the config is loaded.
+- User management, auth etc. This is just raw sessions, but they may be useful if someone is implementing auth
+- Building or maintaining our own driver backends. We will rely on existing libraries for this.
+- Strong consistency, atomic writes
+
+# Detailed Design
+
+The session object is a wrapper around pluggable storage backends. It is lazy-loaded: the contents of the session is fetched from the data store when `get()` is first called, and is `set()` at the end of the request. This means each request is treated as a single transaction, with read-your-writes consistency through the local object. Persistent storage is last-write-wins, with no attempt made to reconcile transactions between multiple requests. Each backend has its own consistency model for the persistent storage, and where this is configurable this will be exposed as part of the driver options.
+
+## API
+
+The session object is available in all Astro contexts, including components, actions, and API endpoints. In components, it is accessed via the global `Astro` object, and in actions and API endpoints it is available on the `context` object. The API is the same in all cases.
+
+Values are serialized and deserialized using [devalue](https://github.com/Rich-Harris/devalue), which is the same library used by content layer and actions. This means that supported types are the same, and include strings, numbers, Dates, Maps, Sets, Arrays and plain objects.
+
+### `Astro.session.get(key: string): Promise<any>`
+
+Returns the value of the given key in the session. If the key does not exist, it returns `undefined`.
+
+### `Astro.session.set(key: string, value: any, options?: { ttl: number }): void`
+
+Sets the value of the given key in the session. The value can be any serializable type.
+
+### `Astro.session.flash(key: string, value: any): void`
+
+Sets the value of the given key in the session. The value can be any serializable type. The value will be deleted after the next request.
```

Comment:
`Astro.locals` **persist** across rewrites and redirect 🤔

Reply:
How does `Astro.locals` persist across requests? If I have a redirect, how would I read the value in the target page?

Reply:
OK, I understand what you meant. I understood "across requests" differently. `Astro.locals` doesn't persist.

I don't understand the usage of `flash()` though. First, it's written that it persists across requests (page navigations), but then it's written that it disappears after the next request (page navigation). It seems like it's similar to `Astro.locals`.

Reply:
The point is that it is disappears *after* the next request. It is available in the next request, but not beyond that. Think of it as a single-use session value: you save it in one request, and read it in the next, and then delete it. The main use case is the one I've described: passing data to the target page of a redirect. Maybe calling it "once" or something rather than flash would be less confusing.

Reply:
How can safety be granted regarding the order of requests? This API feels very brittle, it is enough for a random image request to come between requests to lose the flashed value.. Maybe flash should be constrained by time rather than by requests?  Time is not deterministic either, but it's more deterministic than request order. Or maybe to a route? to access to a key? a combination of timed and the other solutions?

Reply:
It's only invalidated by requests that actually read the session, so random image requests won't affect it.

Reply:
Those can also sneak in, and the worse the network conditions are the more likely it is to happen. I can go into many specific real-world scenarios where this is very likely. Anything with SWR, pings, SSE, any sort of polling, prefetching, navigation, multiple-tabs, timeouts, there are so many reasons for requests to be sent in indeterminate times. This is even more pronounced in authenticated flows since most interactions would check the identity of the requester, these flows are also the most likely to benefit from the session API

Reply:
`flash` is an optional API. You can always use a regular `session.set` if your use case may have extra requests.

Reply:
How is flash implemented in other frameworks that support it?

Reply:
I basically copied the Laravel API. 

>  Sometimes you may wish to store items in the session for the next request. You may do so using the flash method. Data stored in the session using this method will be available immediately and during the subsequent HTTP request. After the subsequent HTTP request, the flashed data will be deleted. Flash data is primarily useful for short-lived status messages: